File: sdks/sdk-python/src/sigilum/retry.py
"""Portable retry helper with bounded backoff for idempotent operations."""

from __future__ import annotations

import random
import time
from typing import Callable, Optional, TypeVar

T = TypeVar("T")

DEFAULT_ATTEMPTS = 3
DEFAULT_BASE_DELAY_SECONDS = 0.1
DEFAULT_MAX_DELAY_SECONDS = 2.0
DEFAULT_JITTER_RATIO = 0.2


def should_retry_http_status(status: int) -> bool:
    return status in (429, 502, 503, 504)


def _jitter_delay(delay_seconds: float, jitter_ratio: float) -> float:
    if delay_seconds <= 0 or jitter_ratio <= 0:
        return max(0.0, delay_seconds)
    window = delay_seconds * jitter_ratio
    if window <= 0:
        return max(0.0, delay_seconds)
    delta = random.uniform(-window, window)
    return max(0.0, delay_seconds + delta)


def retry_with_backoff(
    operation: Callable[[], T],
    *,
    idempotent: bool,
    attempts: int = DEFAULT_ATTEMPTS,
    base_delay_seconds: float = DEFAULT_BASE_DELAY_SECONDS,
    max_delay_seconds: float = DEFAULT_MAX_DELAY_SECONDS,
    jitter_ratio: float = DEFAULT_JITTER_RATIO,
    should_retry_result: Optional[Callable[[T], bool]] = None,
    should_retry_error: Optional[Callable[[Exception], bool]] = None,
    sleep: Callable[[float], None] = time.sleep,
) -> T:
    if attempts < 1:
        attempts = 1
    if not idempotent and attempts > 1:
        raise ValueError("retry_with_backoff requires idempotent=True when attempts > 1")

    delay_seconds = max(0.0, base_delay_seconds)
    max_delay_seconds = max(delay_seconds, max_delay_seconds)
    jitter_ratio = max(0.0, jitter_ratio)
    should_retry_error = should_retry_error or (lambda _: False)

    for attempt in range(1, attempts + 1):
        try:
            result = operation()
            if attempt < attempts and should_retry_result is not None and should_retry_result(result):
                sleep(_jitter_delay(delay_seconds, jitter_ratio))
                delay_seconds = min(max_delay_seconds, delay_seconds * 2)
                continue
            return result
        except Exception as error:  # noqa: BLE001
            if attempt >= attempts or not should_retry_error(error):
                raise
            sleep(_jitter_delay(delay_seconds, jitter_ratio))
            delay_seconds = min(max_delay_seconds, delay_seconds * 2)

    raise RuntimeError("retry_with_backoff exhausted attempts without a terminal result")

File: sdks/sdk-ts/src/retry.ts
export interface RetryWithBackoffOptions<T> {
  idempotent: boolean;
  attempts?: number;
  baseDelayMs?: number;
  maxDelayMs?: number;
  jitterRatio?: number;
  shouldRetryResult?: (result: T) => boolean;
  shouldRetryError?: (error: unknown) => boolean;
  sleep?: (delayMs: number) => Promise<void>;
}

const DEFAULT_ATTEMPTS = 3;
const DEFAULT_BASE_DELAY_MS = 100;
const DEFAULT_MAX_DELAY_MS = 2_000;
const DEFAULT_JITTER_RATIO = 0.2;

function defaultSleep(delayMs: number): Promise<void> {
  return new Promise((resolve) => {
    setTimeout(resolve, Math.max(0, delayMs));
  });
}

function jitterDelay(delayMs: number, jitterRatio: number): number {
  if (delayMs <= 0 || jitterRatio <= 0) {
    return Math.max(0, delayMs);
  }
  const window = Math.floor(delayMs * jitterRatio);
  if (window <= 0) {
    return Math.max(0, delayMs);
  }
  const delta = Math.floor(Math.random() * ((window * 2) + 1)) - window;
  return Math.max(0, delayMs + delta);
}

export function shouldRetryHttpStatus(status: number): boolean {
  return status === 429 || status === 502 || status === 503 || status === 504;
}

export async function retryWithBackoff<T>(
  operation: () => Promise<T>,
  options: RetryWithBackoffOptions<T>,
): Promise<T> {
  const attempts = Math.max(1, options.attempts ?? DEFAULT_ATTEMPTS);
  if (!options.idempotent && attempts > 1) {
    throw new Error("retryWithBackoff requires idempotent=true when attempts > 1");
  }

  const baseDelayMs = Math.max(0, options.baseDelayMs ?? DEFAULT_BASE_DELAY_MS);
  const maxDelayMs = Math.max(baseDelayMs, options.maxDelayMs ?? DEFAULT_MAX_DELAY_MS);
  const jitterRatio = Math.max(0, options.jitterRatio ?? DEFAULT_JITTER_RATIO);
  const shouldRetryResult = options.shouldRetryResult;
  const shouldRetryError = options.shouldRetryError ?? (() => false);
  const sleep = options.sleep ?? defaultSleep;

  let delayMs = baseDelayMs;

  for (let attempt = 1; attempt <= attempts; attempt += 1) {
    try {
      const result = await operation();
      if (attempt < attempts && shouldRetryResult?.(result) === true) {
        await sleep(jitterDelay(delayMs, jitterRatio));
        delayMs = Math.min(maxDelayMs, delayMs * 2);
        continue;
      }
      return result;
    } catch (error) {
      if (attempt >= attempts || !shouldRetryError(error)) {
        throw error;
      }
      await sleep(jitterDelay(delayMs, jitterRatio));
      delayMs = Math.min(maxDelayMs, delayMs * 2);
    }
  }

  throw new Error("retryWithBackoff exhausted attempts without a terminal result");
}
